### ISSUE TITLE
10. RNN을 이용한 Text Classification

### ISSUE THREAD

**Issue by p-jiho** [p-jiho/natural-language#10]
- 텍스트 분류 : 텍스트가 어떤 종류의 범주에 속하는지 구분하는 작업
- 분류 범주가 두 가지 : 이진 분류, 세 가지 이상 : 다중 클래스 분류
- 감성 분석 : 긍정과 부정을 알아보는 것
- 의도 분석 : 텍스트에 담긴 사용자의 의도를 분류

#### 1) 훈련 데이터에 대한 이해
- 지도 학습 : 데이터와 레이블로 이루어져 정답이 있는 문제를 학습한다.

#### 2) 훈련 데이터와 테스트 데이터
- 훈련 데이터 : 훈련하고 학습하는 과정에서 사용할 데이터
- 테스트 데이터 : 훈련이 끝난 후 분류를 시킨 후 실제 답과 맞춰보며 정확도를 계산한다.

#### 3) 단어에 대한 정수 부여
- `Embedding()` : 정수로 변환된 입력에 대해 임베딩 작업을 한다.
- 등장 빈도 순으로 단어를 정렬했을 때의 장점 : 빈도수가 적은 단어 제거

#### 4) RNN으로 분류
- `model.add(SimpleRNN())`
     - `hidden_units` 인자 : 은닉상태의 크기, RNN의 출력 크기
     - `timesteps` 인자 : 각 문서의 단어 수, 시점의 수
     - `input_dim` 인자 : 입력 크기, 임베딩 벡터의 차원

#### 5) RNN의 다-대-일 문제
- 텍스트 분류는 다 대 일 문제이다. : 모든 시점의 입력을 받지만 최종 시점의 은닉 상태만 출력하고 활성화 함수를 통해 답을 고른다.
- 이진 분류일 경우 : 활성화함수는 시그모이드 함수, 손실 함수는 binary_crossentropy를 사용한다.
- 다중 클래스일 경우 : 활성화함수는 소프트맥스 함수, 손실 함수는 categorical_cressentropy를 사용한다.
                                     또한, 클래스가 N개 라면 출력층인 밀집층의 크기는 N이다.

<br></br>

#### 1) 스팸 메일 데이터에 대한 이해
- 파일 다운로드 : https://www.kaggle.com/uciml/sms-spam-collection-dataset
- `urllib.request.urlretrieve` 함수를 이용해 스팸 데이터를 링크로 다운받을 수 있다.
- 전처리
    - `del` 함수를 이용해 불필요한 열을 지운다.
    - `replace` 함수를 이용해 레이블 변환을 한다.
    - `isnull().values.any()` : Null 값을 가진 샘플이 있는지 확인
    - `nunique()` : 중복이 있는지 확인
    - `drop_duplicates` : 중복 제거
- 훈련 데이터와 테스트 데이터로 분리할 때 문제 : 레이블이 불균형한 경우 데이터를 분리할 때 한쪽으로 쏠리지 않도록 고르게 분포되어야한다.
    - 해결 : `train_test_split`의 `stratify`의 인자로 레이블 데이터를 기재하면 된다.
- 정수 인코딩과 토큰화
    - `Tokenizer()` : 토큰화 준비
    - `fit_on_texts` : 단어 집합을 생성하여 정수 인코딩이 된다.
    - `texts_to_sequences` : 토큰화
- 빈도가 많은 순서대로 n개로 제한한다. : `Tokenizer(num_words = n)`
- 패딩 작업 : `pad_sequences`

#### 2) RNN으로 스팸 메일 분류하기
- 모델 생성 : `Sequential()`
- 임베딩 층 생성 : `Embedding(단어집합의 크기, 임베딩 벡터 차원)`
- RNN 수행 : `SimpleRNN(은닉상태의 크기)`
- 출력층  생성 : `Dense(1, activation="sigmoid")`
- p%는검증 데이터로 분류 : `model.fit` 함수의 `validataion_split=p` 인자로 지정
- 정확도 확인 : test 데이터를 토큰화하고 패딩 작업을 거친 후 `evaluate` 함수 적용

<br></br>

#### 1) 로이터 뉴스 데이터에 대한 이해
- `from tensorflow.keras.datasets import reuters`
- 로이터 뉴스 데이터 로드 : `reuters.load_data`
         - `num_words` 인자 : 빈도 수 기준 상위 n개의 단어를 사용. `None`으로 지정하면 전체 단어 사용
         - `test_split` 인자 : 테스트 데이터로 사용할 비율
    -> 토큰화와 정수 인코딩이 끝난 결과를 불러온다.
- `reuterrs.get_word_index` : 단어와 단어에 부여된 인덱스 리턴
- 정수로부터 단어를 알수있도록 key와 value 반대로 저장
    - key와 value로 구분할 수 있는데 실제 저장된 값에 +3을 해야 실제 맵핑 정수가 된다. 즉, 4를 입력해야 1에 대한 단어가 나온다.
    - 0은 패딩을 의미하는 pad, 1은 문장 시작을 의미하는 sos, 2는 OOV를 위한 토큰인 unk로 매핑한다.
    -> 각 값을 `enumerate` 함수를 이용해 인덱스와 값으로 연결을 해준다.

#### 2) LSTM으로 로이터 뉴스 분류하기
- 데이터를 불러오고 `pad_sequences`를 이용해 뉴스 기사에는 패딩 작업을 하고 `to_categorical`을 이용해 레이블에는 원-핫 인코딩을 한다.
- 다중 클래스 분류 문제이므로 소프트맥스 함수, 크로스 엔트로피 함수를 사용한다.
- `model.fit`
    - `validation_data` 인자 : test 데이터를 사용했다. 이 데이터를 학습하지는 않았지만 loss를 계산하는데 사용하였다.
    - `validation_split` 인자 : 검증 데이터를 사용한다.
    - `validation_split`을 사용하는 것이 좋지만 데이터가 작은 경우는 test 데이터로 손실을 구해야한다.

<br></br>

#### 1) IMDB 리뷰 데이터에 대한 이해
- `imdb.load_data()`를 통해 영화 리뷰 데이터를 로드 : 훈련 데이터와 테스트 데이터를 5:5 비율로 미리 구분해서 제공한다.
    -> `test_split` 인자와 같은 것은 제공하지 않는다.
- 레이블의 분포는 동일하게 저장되어있다.
- 로이터 데이터에서 `get_word_index()`를 이용하고 key와 value 값을 바꾼 코드와 거의 동일하게 실행한다.
- 로이터 데이터에서 pad, sos, unk을 지정했던 코드와 거의 동일하게 실행한다.

#### 2) GRU로 IMDB 리뷰 감성 분류하기
- `pad_sequences` 함수를 이용해 패딩을 진행한다.(레이블이 이진 분류이므로 원-핫 인코딩을 할 필요가 없다.)
- 다 대 일 구조의 LSTM을 사용했던 로이터 데이터와 달리 LSTM와 비슷한 다 대 일 구조의 GRU를 사용한다.
- 이진 분류 문제이므로 로지스틱 회귀, 시그모이드 함수, 크로스 엔트로피 함수를 사용한다.
- 출력층 값 : 로이터 데이터에서는 클래스가 46개이므로 46개의 출력 뉴런을 지정했다면 여기는 이진 분류이므로 1개만 지정하엿다.

<br></br>

## 5. 나이브 베이즈 분류기
- 머신러닝의 주요 알고리즘
#### 1) 베이즈 정리를 이용한 분류 메커니즘
- 나이브 베이즈 분류기는 베이즈 정리를 이용해 텍스트 분류를 한다.
- 조건부확률을 보고 확률이 더 큰 것을 선택하는 방식이다.
- 입력은 본문을 단어 토큰화하여 사용한다.
- 단어의 순서는 중요하지 않고 빈도수만을 고려한다.

#### 2) 스팸 메일 분류기
- 문제 : 단 하나의 단어라도 없으면 확률이 0이 되어버린다. -> 지나친 일반화
- 해결 : 라플라스 스무딩 - 각 단어에 대한 확률의 분모와 분자에 전부 숫자를 더하여 0을 방지한다.

#### 3) 뉴스그룹 데이터 분류하기
- 사이킷 런에서 제공하는 20개의 다른 주제를 가진 뉴스 데이터

1. 뉴스그룹 데이터에 대한 이해
- `fetch_20newsgroups`를 이용해 데이터를 받는다.
- 20개의 카테고리가 있고 `target` 속성으로 카테고리 번호를 알고 `target_names`로 제목을 알 수 있다.

2. 나이브 베이즈 분류
- `CountVectorizer` : 입력한 텍스트를 BoW(순서 고려하지 않고 빈도수만 측정한 것)으로 만든다.
     - `fit_transform(data)` : `data`를 BoW로 변환해주는 함수
- 얻은 DTM을 그대로 사용가능하지만 TF-IDF 가중치를 적용하면 성능이 개선될 수 있다.
- `TfidVectorizer` : 사이킷런에서 제공하는 TF-IDF를 계산해주는 함수
     - `fit_transform(dtm)` : DTM 행렬을 넣으면 TF-IDF로 변환해준다.
- `MultinomialNB()` : 사이킷런에서 제공하는 나이브 베이즈 모델 
     - `fit` : 첫번째 인자는 TF-IDF 행렬, 두번째 인자는 레이블이다.
     - `alpha=1.0` : 라플라스 스무딩을 적용하는 인자이다.
- `predict` : 테스트 데이터를 입력하면 예측을 해준다.
- `accuracy_score` : 예측값과 실제값 비교를 해준다. 즉, 정확도를 구해준다.

<br></br>

## 6. 네이버 영화 리뷰 감성 분류하기
- 데이터 링크 받아오기 : https://github.com/e9t/nsmc/

#### 1) 데이터 로드
- 받아온 링크를 통해 `urllib.request.urlretrieve` 함수를 이용해 train과 test를 다운로드한다.

#### 2) 데이터 정제
- `nunique()` : 중복 유무 확인
- `drop_duplicated` : 중복 데이터 제거
- `isnull().values.any()` : Null 값을 가진 샘플 확인
- `dropna(how="any")` : Null 값이 존재하는 행 제거
- `str.replace` : 정규표현식을 이용해 제거

#### 3) 토큰화
- KoNLPy의 `Okt()` : 토큰화를 위한 형태소 분석기
     - `stem=True` 인자 : 일정한 수준의 정규화를 해준다.
- `okt.morphs` : 토큰화 진행
- 지정한 불용어 리스트를 `if` 문과 `for` 문을 이용해 제거한다.

#### 4) 정수 인코딩
- `Tokenizer()`
- `fit_on_texts` : 단어 집합 생성
- `word_index` : 정수 인코딩 결과 확인
- `Tokenizer(vocab_size)` : 등장 빈도수가 2 이하인 단어 제거, `vacab_size`를 통해 최대 갯수를 지정해 빈도수가 높은 순으로 자른다. 즉, `vocab_size`를 등장 빈도수가 3 이상인 단어 수로 지정한다.
- `texts_to_sequences` : 단어의 시퀀스를 숫자로 변환

5. 빈 샘플 제거
- 빈도 수가 낮은 단어를 제거했으므로 몇몇은 빈 샘플이 되었다.
- 샘플의 길이가 0인 샘플들의 인덱스를 구하고 `np.delete`를 이용해 제거한다.

6. 패딩
- 서로 다른 길이의 샘플을 동일한 길이로 맞춰주는 작업
- 최대 길이로 하면 공간이 낭비되므로 많은 데이터의 길이를 넘는 수 중에서 가장 최소화된 길이로 지정을 한다.

#### 2) LSTM으로 네어버 영화 리뷰 감성 분류
- 이진 분류이므로 시그모이드함수, 크로스 엔트로피를 사용한다.
- 모델 생성 : `Sequential()`
- 층 생성 : `Embedding`, `LSTM`, `Dense`

#### 3) 리유 예측해보기
- `model.predict` : 예측값을 얻는다.
- 동일하게 형태소 분석기, 정규표현식을 이용한 전처리, 불용어 제거, 정수 인코딩, 패딩을 한 정제된 데이터로 예측을 한다.

<br></br>

#### 1. 네이버 쇼핑 리뷰 데이터에 대한 이해와 전처리
- 링크 : https://github.com/bab2min/corpus/tree/master/sentiment

1. 데이터 로드하기
- `urlib.request.urlretrieve` 함수를 이용해 naver_shopping.txt를 다운로드한다.

2. 훈련 데이터와 테스트 데이터 분리하기
- 평점이 4,5 이면 1, 평점이 1,2이면 0으로 레이블을 부여한다.
- 중복 제거 : `drop_duplicates`
- 데이터 분리 : `train_test_split`, `test_size` 인자를 이용해 테스트 데이터 비율을 지정한다.

3. 데이터 정제
- `str.replace` : 정규표현식을 이용해 한글과 공백을 제외하고는 전부 제거한다.

4. 토큰화
- 형태소 분석기인 Mecab을 사용해 토큰화 작업 수행
- `apply(mecab.morphs)`를 이용해 토큰화 작업을 한다.
- `apply`와 `lambda`를 이용해 불용어 제거 작업도 한다.

5. 단어 확인
- 긍정, 부정 리뷰에 대한 단어 리스트를 저장한다. : `np.hstack`
- 각 단어의 빈도수 카운트 : `Counter`

6. 정수 인코딩
- 등장 빈도가 적은 단어들은 제거하는 것이 좋다.
- `Tokenizer(vocab_size, oov_token="OOV")` : `vocab_size` 보다 큰 숫자가 부여된 단어는 OOV로 변환한다.
- `fit_on_texts` : 리스트 형태로 문자를 변환
- `texts_to_sequences` : 정수 인코딩을 한다.

7. 패딩
- 서로 다른 길이의 샘플들의 길이를 동일하게 맞춘다.
- `pad_sequences` : `maxlen` 인자를 이용해 길이를 정할 수 있다. 대부분의 샘플들이 온전히 보전되는 선에서 값을 정하는게 좋다.

#### 3) GRU로 네이버 쇼핑 리뷰 감성 분류
- 다 대 일 구조의 GRU를 사용한다.
- 이진 분류 문제이므로 시그모이드 함수, 크로스 엔트로피 함수를 사용한다.
- 코드는 전의 과정들과 거의 동일하다.

#### 4) 리뷰 예측하기
- 학습할 데이터를 전과 동일하게 전처리한다.
- `predict` : 예측을 한다.

<br></br>

#### 1) BiLSTM을 텍스트 분류에 사용하기
- 순방향 LSTM과 역방향LSTM을 모두 고려해 출력층에서 두 가지 정보를 모두 사용한다.
- 다 대 일 BiLSTM을 사용하면 마지막 시점에서 역방향 LSTM은 유용한 정보를 가지지 않는다.
   -> `return_sequences=False`를 탱새 순방향은 마지막 시점, 역방향은 첫번째 시점의 은닉상태를 반환한다.

#### 2) 스팀 리뷰 데이터에 대한 이해와 전처리
- 링크 : https://github.com/bab2min/corpus/tree/master/sentiment

1. 데이터 로드
- steam.txt 데이터를 `urllib.request.urlretrieve`를 이용해 다운받는다.
- `pd.read_table`을 통해 데이터를 가져올 수 있고 `names` 인자를 이용해 열 이름 지정이 가능하다.
- `nunique()` 함수를 이용해 중복없이 샘플 수를 카운트한 후 중복 샘플이 있으면 `drop_duplicates`를 이용해 없애준다.

2. 훈련 데이터와 테스트 데이터 분리
- `train_test_split`을 이용하여 분리한다.

3. 레이블 분포 확인
- label 열을 `value_counts()` 함수를 이용해 label 별 수를 계산한다.

4. 데이터 정제
- `str.replace` : 정규표현식에 해당하는 샘플을 제외하고는 제거한다.

5. 토큰화
- `Mecab()`을 이용해 `apply(mecab.morphs)`를 이용해 토큰화 작업을 한다.
- 불용어 리스트를 만들어 `apply` 함수와 `lambda`를 이용해 불용어를 없애준다.

6. 단어와 길이 분포 확인
- `np.hstack`을 이용해 리스트로 저장을 해준다.
- `Counter`을 이용해 빈도수를 구한다.
- 길이에 대한 분포는 `map(lambda x : len(x))`을 이용해서 구한다.

7. 정수 인코딩
- `Tokenizer()`을 이용하고 `fit_on_texts`를 이용해 단어 집합을 만든다.
- `tokenizer.word_counts.items()`를 이용해 key와 value를 받아 value가 2보다 작으면 빈도수가 1인 값이다. 
    -> 빈도수가 1인 단어를 배제한다
- 0번 패딩 토큰, 1번 OOV 토큰을 고려하여 단어 집합의 최대 크기를 제한한다. : `Tokenizer(vocab_size, oov_token="OOV")`
- `fit_on_texts`와 `texts_to_sequences`를 이용해 정수 인코딩을 한다.

8. 패딩
- 패딩을 진행할 적절한 값을 구한후 `pad_sequences`를 이용해 패딩을 진행한다.

#### 4) BiLSTM으로 스팀 리뷰 감성 분류하기
- 영화 리뷰, 쇼핑 리뷰 분류와 비슷하다.
- 양방향 LSTM을 사용했으므로 `Bidirectional(LSTM(hidden_units)))`을 사용한다.

#### 5) 리뷰 예측하기
- 데이터에 전처리 과정을 동일하게 적용한 후 `predict` 함수를 이용해 예측한다.

---

- 참고
>> 딥 러닝을 이용한 자연어 처리 입문, 유원준, 안상준
      * 위 글 내용은 이 책의 내용을 공부하여 정리한 것입니다.
